### ISSUE TITLE
Using Blaze Components

### ISSUE THREAD

**Issue by SachaG** [meteor-utilities/menu#1]
I've noticed some performance issues when dealing with medium (>100) numbers of nested menu items. I was wondering if maybe refactoring this package to use [Blaze Components](https://github.com/peerlibrary/meteor-blaze-components) might help? 

Would love to get @mitar's feedback, and also get more people involved in using/maintaining this package.

**Comment by mitar** [comment]
Oho, that is an interesting package. And I like animations.

What type of performance issues you have? It would be good to measure first things. Is it reactivity related?

But would definitely be a good use case otherwise for Blaze Components. It has a nice component structure, and it would be great to try out the animation hooks for this.

**Comment by mitar** [comment]
Can you make an example where it is slow?

**Comment by SachaG** [comment]
I was actually talking about it with @lorensr, maybe he can describe the issue as well. I have a site with about 140 menu items, and when I trigger a route that renders the menu with all of them at once, it can get pretty slow (anywhere from a couple seconds to >30 seconds if you go back and forth a lot). 

Here's a video: http://cl.ly/dZUl

I can also create a reproduction in a separate repo if you'd like. 

What I ended up doing was only rendering the first level of the menu until you expand it. But that meant giving up on animations… (I was doing them via jQuery, and this requires having the thing you want to show already rendered in the DOM).

**Comment by lorensr** [comment]
In Telescope, using `menu` on `/submit`. With or without `expandLevel="all"`:
- on `/page/about`
- go to `/submit`, takes 3s
- go to `/page/about`
- go to `/submit`, takes 10s

Continues increasing time to both load `/submit` and the time to tear down `/submit` template when moving to `/page/about`. Compare to time it takes to go between two routes that don't have `menu`, eg `/users/yourname` and `/page/about`: 1s, non-increasing.

Timeline profiling video:

https://youtu.be/bRRqwJ4I0nA

I guess the long lines of repeating `Blaze._materializeDOM` and alternating `DOMRange.attach` / `DOMRange._insert` are related to the use of nested loops of `Template.dynamic`. A lot of event attaching, maybe could only have one set of event handlers on the parent div. Don't know why it takes longer on subsequent renders – something must not be cleaned up?

**Comment by lorensr** [comment]
@SachaG You can still have animations with Blaze! I use `percolate:momentum`, with `{{#momentum plugin='css'}}` – pretty easy to use :)

**Comment by SachaG** [comment]
You're right, I just haven't gotten around to adding them back in yet. But I figure it's probably best to solve any performance issue first before I add more variables to the mix :)

**Comment by mitar** [comment]
> I can also create a reproduction in a separate repo if you'd like. 

Are you sure it was not a data loading issue? So that you were trying to render the menu before all data was there from the server? Or is menu hard-coded?

On animations, Blaze Components also have animations hook.

OK, so would you be interested in me trying to port this to Blaze Components and then you can evaluate?

But first it would be good to find some static reproduction. So something where data is not loading from the server, and maybe two routes you move between.

**Comment by mitar** [comment]
Hm, it would be interesting to see just how many DOM elements and template instances are there. It might be simply a lot of those?

**Comment by lorensr** [comment]
How do you count template instances?

**Comment by mitar** [comment]
You put increasing a global variable in `onCreated` of all of them? ;-)

**Comment by mitar** [comment]
You can also measure timestamp of each, and then also rendering time (difference between calling `onCreated` and `onRendered`).

**Comment by lorensr** [comment]
Oh, hoped you could enumerate on console of a rendered page

**Comment by mitar** [comment]
Yes, you can, you access that variable in the console. ;-)

**Comment by lorensr** [comment]
I mean like I can see how many packages by `Object.keys(Package).length`. I want something like `Blaze._templateInstances.length`, but can't find it.

**Comment by mitar** [comment]
Yes, instrumenting Blaze to count that might potentially slow things down, so I do think it is good that it is not there. But one can probably instrument it probably quite easy by extending the template instance prototype? Or even View prototype.

**Comment by lorensr** [comment]
![image](https://cloud.githubusercontent.com/assets/251288/10596544/5551e972-76b3-11e5-9604-7bac39e2db11.png)

20k nodes first time /submit is rendered, 24k second time, 29k third time (not shown).

**Comment by lorensr** [comment]
Here's a method of counting: 

http://stackoverflow.com/questions/33227164/how-do-you-get-all-instances-of-a-template/33235234#33235234

```
> findAllTemplateInstances('Template.menuItem').length
286
```
